In [1]:
import os
import pandas as pd
import numpy as np

In [2]:
RAW_PATH = "../data/raw"
PROCESSED_PATH = "../data/processed"

In [3]:
customers = pd.read_csv(os.path.join(RAW_PATH, "olist_customers_dataset.csv"))

products = pd.read_csv(os.path.join(RAW_PATH, "olist_products_dataset.csv"))

order_items = pd.read_csv(os.path.join(RAW_PATH, "olist_order_items_dataset.csv"))

payments = pd.read_csv(os.path.join(RAW_PATH, "olist_order_payments_dataset.csv"))

reviews = pd.read_csv(os.path.join(RAW_PATH, "olist_order_reviews_dataset.csv"))

sellers = pd.read_csv(os.path.join(RAW_PATH, "olist_sellers_dataset.csv"))

category_translation = pd.read_csv(
    os.path.join(RAW_PATH, "product_category_name_translation.csv")
)

In [11]:
datasets = {
    "Customers": customers,
    "Products": products,
    "Order Items": order_items,
    "Payments": payments,
    "Reviews": reviews,
    "Sellers": sellers,
    "Category Translation": category_translation,
}

In [19]:
for name, df in datasets.items():
    print("=" * 70)
    print(name.upper())
    print("=" * 70)

    print("Shape:", df.shape)

    print("\nColumns:")
    print(df.columns.tolist())

    print("\nData Types:")
    print(df.dtypes)

    print("\nMissing Values:")
    print(df.isnull().sum())

    print("\nDuplicate Rows:", df.duplicated().sum())

CUSTOMERS
Shape: (99441, 5)

Columns:
['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

Data Types:
customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object

Missing Values:
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

Duplicate Rows: 0
PRODUCTS
Shape: (32951, 9)

Columns:
['product_id', 'product_category_name', 'product_name_lenght', 'product_description_lenght', 'product_photos_qty', 'product_weight_g', 'product_length_cm', 'product_height_cm', 'product_width_cm']

Data Types:
product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g       

In [20]:
from IPython.display import display

for name, df in datasets.items():
    print(f"\n{name}")
    display(df.head())


Customers


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP



Products


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0



Order Items


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14



Payments


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45



Reviews


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53



Sellers


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP



Category Translation


,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto
3,cama_mesa_banho,bed_bath_table
4,moveis_decoracao,furniture_decor


In [22]:
customers_raw = pd.read_csv("../data/raw/olist_customers_dataset.csv")
customers_processed = pd.read_csv("../data/processed/customers_clean.csv")

print(customers_raw.shape)
print(customers_processed.shape)

print(customers_raw.equals(customers_processed))

(99441, 5)
(99441, 5)
True


In [23]:
customers.head() #inspect the first few rows of the customers dataframe
customers.info() # check the structure of the customers dataframe
customers.isnull().sum() # check for missing values in the customers dataframe  
customers.duplicated().sum() # check for duplicate rows in the customers dataframe



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


np.int64(0)

In [26]:
customers["customer_id"].duplicated().sum() # check for duplicate customer IDs in the customers dataframe


np.int64(0)

In [27]:
customers["customer_unique_id"].duplicated().sum() # check for duplicate unique customer IDs in the customers dataframe

np.int64(3345)

In [28]:
customers.select_dtypes(include="object").apply(
    lambda col: (col.str.strip() == "").sum()
)

customer_id           0
customer_unique_id    0
customer_city         0
customer_state        0
dtype: int64

In [29]:
customers["customer_state"].value_counts().sort_index()

customer_state
AC       81
AL      413
AM      148
AP       68
BA     3380
CE     1336
DF     2140
ES     2033
GO     2020
MA      747
MG    11635
MS      715
MT      907
PA      975
PB      536
PE     1652
PI      495
PR     5045
RJ    12852
RN      485
RO      253
RR       46
RS     5466
SC     3637
SE      350
SP    41746
TO      280
Name: count, dtype: int64

In [30]:
text_columns = ["customer_city", "customer_state"]

for col in text_columns:
    customers[col] = customers[col].str.strip()

In [31]:
customers.info()
customers.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [32]:
customers.to_csv(
    os.path.join(PROCESSED_PATH, "customers_clean.csv"),
    index=False
)

In [33]:
customers.info()
customers.isnull().sum()
customers.duplicated().sum()
customers["customer_id"].duplicated().sum()
customers["customer_unique_id"].duplicated().sum()
customers["customer_state"].value_counts().sort_index()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   customer_id               99441 non-null  object
 1   customer_unique_id        99441 non-null  object
 2   customer_zip_code_prefix  99441 non-null  int64 
 3   customer_city             99441 non-null  object
 4   customer_state            99441 non-null  object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


customer_state
AC       81
AL      413
AM      148
AP       68
BA     3380
CE     1336
DF     2140
ES     2033
GO     2020
MA      747
MG    11635
MS      715
MT      907
PA      975
PB      536
PE     1652
PI      495
PR     5045
RJ    12852
RN      485
RO      253
RR       46
RS     5466
SC     3637
SE      350
SP    41746
TO      280
Name: count, dtype: int64

In [37]:
customers.isnull().sum()



customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

## Customers Dataset Cleaning Summary

### Validation Results
- No missing values found.
- No duplicate rows found.
- `customer_id` is unique.
- `customer_unique_id` contains expected duplicates because a customer can place multiple orders.
- Data types are correct.
- State abbreviations are valid.
- Leading/trailing whitespace removed from text columns.

### Outcome
The dataset required minimal cleaning and is ready for downstream merging.

In [38]:
# ============================================================
# Products Dataset Cleaning
# ============================================================

print("=" * 60)
print("PRODUCTS DATASET")
print("=" * 60)

# -------------------------
# Basic Information
# -------------------------
print(f"Shape: {products.shape}")

print("\nData Types:")
print(products.dtypes)

# -------------------------
# Missing Values
# -------------------------
print("\nMissing Values:")
print(products.isnull().sum())

# -------------------------
# Duplicate Checks
# -------------------------
print("\nDuplicate Rows:", products.duplicated().sum())
print("Duplicate product_id:",
      products["product_id"].duplicated().sum())

# -------------------------
# Blank String Check
# -------------------------
print("\nBlank Strings:")
print(
    products.select_dtypes(include="object").apply(
        lambda col: (col.str.strip() == "").sum()
    )
)

# -------------------------
# Standardize Text Columns
# -------------------------
text_columns = ["product_category_name"]

for col in text_columns:
    products[col] = products[col].str.strip()

# -------------------------
# Final Validation
# -------------------------
print("\nFinal Shape:", products.shape)

# -------------------------
# Save Clean Dataset
# -------------------------
products.to_csv(
    os.path.join(PROCESSED_PATH, "products_clean.csv"),
    index=False
)

print("\n✅ Products dataset cleaned and saved successfully.")

PRODUCTS DATASET
Shape: (32951, 9)

Data Types:
product_id                     object
product_category_name          object
product_name_lenght           float64
product_description_lenght    float64
product_photos_qty            float64
product_weight_g              float64
product_length_cm             float64
product_height_cm             float64
product_width_cm              float64
dtype: object

Missing Values:
product_id                      0
product_category_name         610
product_name_lenght           610
product_description_lenght    610
product_photos_qty            610
product_weight_g                2
product_length_cm               2
product_height_cm               2
product_width_cm                2
dtype: int64

Duplicate Rows: 0
Duplicate product_id: 0

Blank Strings:
product_id               0
product_category_name    0
dtype: int64

Final Shape: (32951, 9)

✅ Products dataset cleaned and saved successfully.


In [39]:
# Remove products with missing product information
products = products.dropna(
    subset=[
        "product_category_name",
        "product_name_lenght",
        "product_description_lenght",
        "product_photos_qty"
    ]
)

# Remove products missing physical dimensions
products = products.dropna(
    subset=[
        "product_weight_g",
        "product_length_cm",
        "product_height_cm",
        "product_width_cm"
    ]
)

In [40]:
print(products.isnull().sum())
print(products.shape)

product_id                    0
product_category_name         0
product_name_lenght           0
product_description_lenght    0
product_photos_qty            0
product_weight_g              0
product_length_cm             0
product_height_cm             0
product_width_cm              0
dtype: int64
(32340, 9)


In [41]:
products.to_csv(
    os.path.join(PROCESSED_PATH, "products_clean.csv"),
    index=False
)

## Products Dataset Cleaning Summary

### Validation
- No duplicate rows
- No duplicate product IDs
- No blank strings

### Cleaning Performed
- Removed 610 products with missing category and descriptive attributes.
- Removed 2 products with missing physical dimensions.

### Outcome
The dataset now contains only complete product records and is ready for downstream merging and analysis.

1. Validate
2. Analyze results
3. Decide cleaning strategy
4. Apply transformations
5. Validate again
6. Save

In [44]:
# ============================================================
# Sellers Dataset Cleaning
# ============================================================

print("=" * 60)
print("SELLERS DATASET")
print("=" * 60)

# -------------------------
# Basic Information
# -------------------------
print(f"Shape: {sellers.shape}")

print("\nData Types:")
print(sellers.dtypes)

# -------------------------
# Missing Values
# -------------------------
print("\nMissing Values:")
print(sellers.isnull().sum())

# -------------------------
# Duplicate Checks
# -------------------------
print("\nDuplicate Rows:", sellers.duplicated().sum())
print("Duplicate seller_id:",
      sellers["seller_id"].duplicated().sum())

# -------------------------
# Blank String Check
# -------------------------
print("\nBlank Strings:")
print(
    sellers.select_dtypes(include="object").apply(
        lambda col: (col.str.strip() == "").sum()
    )
)

# -------------------------
# Standardize Text Columns
# -------------------------
text_columns = ["seller_city", "seller_state"]

for col in text_columns:
    sellers[col] = sellers[col].str.strip()

# -------------------------
# Validation After Cleaning
# -------------------------
print("\nRemaining Missing Values:")
print(sellers.isnull().sum())

print("\nFinal Shape:", sellers.shape)
# -------------------------
# Save Clean Dataset
# -------------------------
sellers.to_csv(
    os.path.join(PROCESSED_PATH, "sellers_clean.csv"),
    index=False
)

print("\n✅ Sellers dataset cleaned and saved successfully.")




SELLERS DATASET
Shape: (3095, 4)

Data Types:
seller_id                 object
seller_zip_code_prefix     int64
seller_city               object
seller_state              object
dtype: object

Missing Values:
seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

Duplicate Rows: 0
Duplicate seller_id: 0

Blank Strings:
seller_id       0
seller_city     0
seller_state    0
dtype: int64

Remaining Missing Values:
seller_id                 0
seller_zip_code_prefix    0
seller_city               0
seller_state              0
dtype: int64

Final Shape: (3095, 4)

✅ Sellers dataset cleaned and saved successfully.


In [45]:
# ============================================================
# Category Translation Dataset Cleaning
# ============================================================

print("=" * 60)
print("CATEGORY TRANSLATION DATASET")
print("=" * 60)

# -------------------------
# Basic Information
# -------------------------
print(f"Shape: {category_translation.shape}")

print("\nData Types:")
print(category_translation.dtypes)

# -------------------------
# Missing Values
# -------------------------
print("\nMissing Values:")
print(category_translation.isnull().sum())

# -------------------------
# Duplicate Checks
# -------------------------
print("\nDuplicate Rows:", category_translation.duplicated().sum())

print(
    "Duplicate Portuguese Categories:",
    category_translation["product_category_name"].duplicated().sum()
)

print(
    "Duplicate English Categories:",
    category_translation["product_category_name_english"].duplicated().sum()
)

# -------------------------
# Blank String Check
# -------------------------
print("\nBlank Strings:")
print(
    category_translation.select_dtypes(include="object").apply(
        lambda col: (col.str.strip() == "").sum()
    )
)

# -------------------------
# Standardize Text Columns
# -------------------------
text_columns = category_translation.select_dtypes(include="object").columns

for col in text_columns:
    category_translation[col] = category_translation[col].str.strip()

# -------------------------
# Final Validation
# -------------------------
print("\nRemaining Missing Values:")
print(category_translation.isnull().sum())

print("\nFinal Shape:", category_translation.shape)

# -------------------------
# Save Clean Dataset
# -------------------------
category_translation.to_csv(
    os.path.join(PROCESSED_PATH, "category_translation_clean.csv"),
    index=False
)

print("\n✅ Category Translation dataset cleaned and saved successfully.")

CATEGORY TRANSLATION DATASET
Shape: (71, 2)

Data Types:
product_category_name            object
product_category_name_english    object
dtype: object

Missing Values:
product_category_name            0
product_category_name_english    0
dtype: int64

Duplicate Rows: 0
Duplicate Portuguese Categories: 0
Duplicate English Categories: 0

Blank Strings:
product_category_name            0
product_category_name_english    0
dtype: int64

Remaining Missing Values:
product_category_name            0
product_category_name_english    0
dtype: int64

Final Shape: (71, 2)

✅ Category Translation dataset cleaned and saved successfully.


In [46]:
# ============================================================
# Order Items Dataset Cleaning
# ============================================================

print("=" * 60)
print("ORDER ITEMS DATASET")
print("=" * 60)

# -------------------------
# Basic Information
# -------------------------
print(f"Shape: {order_items.shape}")

print("\nData Types:")
print(order_items.dtypes)

# -------------------------
# Missing Values
# -------------------------
print("\nMissing Values:")
print(order_items.isnull().sum())

# -------------------------
# Duplicate Checks
# -------------------------
print("\nDuplicate Rows:", order_items.duplicated().sum())
print("Duplicate order_id + order_item_id:",
      order_items.duplicated(subset=["order_id", "order_item_id"]).sum())

# -------------------------
# Blank String Check
# -------------------------
print("\nBlank Strings:")
print(
    order_items.select_dtypes(include="object").apply(
        lambda col: (col.str.strip() == "").sum()
    )
)

# -------------------------
# Data Type Conversion
# -------------------------
order_items["shipping_limit_date"] = pd.to_datetime(
    order_items["shipping_limit_date"]
)

# -------------------------
# Standardize Text Columns
# -------------------------
text_columns = ["order_id", "product_id", "seller_id"]

for col in text_columns:
    order_items[col] = order_items[col].str.strip()

# -------------------------
# Final Validation
# -------------------------
print("\nRemaining Missing Values:")
print(order_items.isnull().sum())

print("\nFinal Shape:", order_items.shape)

# -------------------------
# Save
# -------------------------
order_items.to_csv(
    os.path.join(PROCESSED_PATH, "order_items_clean.csv"),
    index=False
)

print("\n✅ Order Items dataset cleaned and saved successfully.")

ORDER ITEMS DATASET
Shape: (112650, 7)

Data Types:
order_id                object
order_item_id            int64
product_id              object
seller_id               object
shipping_limit_date     object
price                  float64
freight_value          float64
dtype: object

Missing Values:
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

Duplicate Rows: 0
Duplicate order_id + order_item_id: 0

Blank Strings:
order_id               0
product_id             0
seller_id              0
shipping_limit_date    0
dtype: int64

Remaining Missing Values:
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64

Final Shape: (112650, 7)

✅ Order Items dataset cleaned and saved successfully.


In [47]:
# ============================================================
# Payments Dataset Cleaning
# ============================================================

print("=" * 60)
print("PAYMENTS DATASET")
print("=" * 60)

# -------------------------
# Basic Information
# -------------------------
print(f"Shape: {payments.shape}")

print("\nData Types:")
print(payments.dtypes)

# -------------------------
# Missing Values
# -------------------------
print("\nMissing Values:")
print(payments.isnull().sum())

# -------------------------
# Duplicate Checks
# -------------------------
print("\nDuplicate Rows:", payments.duplicated().sum())

# -------------------------
# Blank String Check
# -------------------------
print("\nBlank Strings:")
print(
    payments.select_dtypes(include="object").apply(
        lambda col: (col.str.strip() == "").sum()
    )
)

# -------------------------
# Standardize Text Columns
# -------------------------
text_columns = ["order_id", "payment_type"]

for col in text_columns:
    payments[col] = payments[col].str.strip()

# -------------------------
# Final Validation
# -------------------------
print("\nRemaining Missing Values:")
print(payments.isnull().sum())

print("\nFinal Shape:", payments.shape)

# -------------------------
# Save
# -------------------------
payments.to_csv(
    os.path.join(PROCESSED_PATH, "payments_clean.csv"),
    index=False
)

print("\n✅ Payments dataset cleaned and saved successfully.")

PAYMENTS DATASET
Shape: (103886, 5)

Data Types:
order_id                 object
payment_sequential        int64
payment_type             object
payment_installments      int64
payment_value           float64
dtype: object

Missing Values:
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Duplicate Rows: 0

Blank Strings:
order_id        0
payment_type    0
dtype: int64

Remaining Missing Values:
order_id                0
payment_sequential      0
payment_type            0
payment_installments    0
payment_value           0
dtype: int64

Final Shape: (103886, 5)

✅ Payments dataset cleaned and saved successfully.


In [48]:
# ============================================================
# Reviews Dataset Cleaning
# ============================================================

print("=" * 60)
print("REVIEWS DATASET")
print("=" * 60)

# -------------------------
# Basic Information
# -------------------------
print(f"Shape: {reviews.shape}")

print("\nData Types:")
print(reviews.dtypes)

# -------------------------
# Missing Values
# -------------------------
print("\nMissing Values:")
print(reviews.isnull().sum())

# -------------------------
# Duplicate Checks
# -------------------------
print("\nDuplicate Rows:", reviews.duplicated().sum())
print("Duplicate review_id:",
      reviews["review_id"].duplicated().sum())

# -------------------------
# Blank String Check
# -------------------------
print("\nBlank Strings:")
print(
    reviews.select_dtypes(include="object").apply(
        lambda col: (col.str.strip() == "").sum()
    )
)

# -------------------------
# Data Type Conversion
# -------------------------
date_columns = [
    "review_creation_date",
    "review_answer_timestamp"
]

for col in date_columns:
    reviews[col] = pd.to_datetime(reviews[col])

# -------------------------
# Standardize Text Columns
# -------------------------
text_columns = [
    "review_id",
    "order_id",
    "review_comment_title",
    "review_comment_message"
]

for col in text_columns:
    reviews[col] = reviews[col].fillna("").str.strip()

# -------------------------
# Final Validation
# -------------------------
print("\nRemaining Missing Values:")
print(reviews.isnull().sum())

print("\nFinal Shape:", reviews.shape)

# -------------------------
# Save
# -------------------------
reviews.to_csv(
    os.path.join(PROCESSED_PATH, "reviews_clean.csv"),
    index=False
)

print("\n✅ Reviews dataset cleaned and saved successfully.")

REVIEWS DATASET
Shape: (99224, 7)

Data Types:
review_id                  object
order_id                   object
review_score                int64
review_comment_title       object
review_comment_message     object
review_creation_date       object
review_answer_timestamp    object
dtype: object

Missing Values:
review_id                      0
order_id                       0
review_score                   0
review_comment_title       87656
review_comment_message     58247
review_creation_date           0
review_answer_timestamp        0
dtype: int64

Duplicate Rows: 0
Duplicate review_id: 814

Blank Strings:
review_id                   0
order_id                    0
review_comment_title        2
review_comment_message     27
review_creation_date        0
review_answer_timestamp     0
dtype: int64

Remaining Missing Values:
review_id                  0
order_id                   0
review_score               0
review_comment_title       0
review_comment_message     0
review_creation

In [50]:
orders = pd.read_csv(
    os.path.join(PROCESSED_PATH, "orders_clean.csv"),
    parse_dates=[
        "order_purchase_timestamp",
        "order_approved_at",
        "order_delivered_carrier_date",
        "order_delivered_customer_date",
        "order_estimated_delivery_date"
    ]
)

In [52]:
# ============================================================
# Final Data Quality Summary
# ============================================================

summary = {
    "Dataset": [
        "Orders",
        "Customers",
        "Products",
        "Sellers",
        "Category Translation",
        "Order Items",
        "Payments",
        "Reviews"
    ],
    "Rows": [
        len(orders),
        len(customers),
        len(products),
        len(sellers),
        len(category_translation),
        len(order_items),
        len(payments),
        len(reviews)
    ],
    "Columns": [
        orders.shape[1],
        customers.shape[1],
        products.shape[1],
        sellers.shape[1],
        category_translation.shape[1],
        order_items.shape[1],
        payments.shape[1],
        reviews.shape[1]
    ]
}

summary_df = pd.DataFrame(summary)

print("=" * 60)
print("FINAL DATASET SUMMARY")
print("=" * 60)

display(summary_df)

FINAL DATASET SUMMARY


,Dataset,Rows,Columns
0,Orders,99441,13
1,Customers,99441,5
2,Products,32340,9
3,Sellers,3095,4
4,Category Translation,71,2
5,Order Items,112650,7
6,Payments,103886,5
7,Reviews,99224,7
